tracing: trait bound error when calling `Span::follows_from`

When attempting to write:

```rust
fn main() {
    let span = span!(Level::INFO, "hello!");
    let curr = Span::current();
    span.follows_from(curr.id());
}
```

Rust fails with the following error:

```
error[E0277]: the trait bound `std::option::Option<&'a tracing_core::span::Id>: std::convert::From<std::option::Option<tracing_core::span::Id>>` is not satisfied
 --> src/main.rs:6:10
  |
6 |     span.follows_from(curr.id());
  |          ^^^^^^^^^^^^ the trait `std::convert::From<std::option::Option<tracing_core::span::Id>>` is not implemented for `std::option::Option<&'a tracing_core::span::Id>`
  |
  = help: the following implementations were found:
            <std::option::Option<&'a T> as std::convert::From<&'a std::option::Option<T>>>
            <std::option::Option<&'a mut T> as std::convert::From<&'a mut std::option::Option<T>>>
            <std::option::Option<T> as std::convert::From<T>>
  = note: required because of the requirements on the impl of `for<'a> std::convert::Into<std::option::Option<&'a tracing_core::span::Id>>` for `std::option::Option<tracing_core::span::Id>`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: could not compile `tracing-reproduction`.

To learn more, run the command again with --verbose.
```

Here's a minimal reproduction: https://github.com/davidbarsky/tracing-reproduction. The only dependency is `tracing-0.1.10`.

--- 

As a workaround, I wrote the following instead:

```rust
dispatcher::get_default(|dispatch| {
    let id = span.id().expect("Missing ID; this is a bug");
    if let Some(current) = dispatch.current_span().id() {
         dispatch.record_follows_from(&id, current)
    }
});
```

--- a/tracing/src/span.rs
+++ b/tracing/src/span.rs
@@ -742,7 +742,7 @@ impl Span {
     /// # }
     /// ```
     ///
-    /// Setting a `follows_from` relationship through a `&Span`:
+    /// Setting a `follows_from` relationship with a `Span` reference:
     /// ```
     /// # use tracing::{span, Id, Level, Span};
     /// # fn main() {

--- a/tracing/src/span.rs
+++ b/tracing/src/span.rs
@@ -721,7 +721,7 @@ impl Span {
     /// If this span is disabled, or the resulting follows-from relationship
     /// would be invalid, this function will do nothing.
     ///
-    /// # Example
+    /// # Examples
     ///
     /// Setting a `follows_from` relationship through an `Option<Id>`:
     /// ```

--- a/tracing/src/span.rs
+++ b/tracing/src/span.rs
@@ -733,7 +733,7 @@ impl Span {
     /// # }
     /// ```
     ///
-    /// Setting a `follows_from` relationship with a `Span::current()`:
+    /// Setting a `follows_from` relationship with the current span:
     /// ```
     /// # use tracing::{span, Id, Level, Span};
     /// # fn main() {

--- a/tracing/src/span.rs
+++ b/tracing/src/span.rs
@@ -723,12 +723,32 @@ impl Span {
     ///
     /// # Example
     ///
+    /// Setting a `follows_from` relationship through an `Option<Id>`:
     /// ```
     /// # use tracing::{span, Id, Level, Span};
     /// # fn main() {
     /// let span = span!(Level::INFO, "hello!");
     /// let curr = Span::current();
     /// span.follows_from(curr.id());
+    /// # }
+    /// ```
+    ///
+    /// Setting a `follows_from` relationship through a `Span`:
+    /// ```
+    /// # use tracing::{span, Id, Level, Span};
+    /// # fn main() {
+    /// let span = span!(Level::INFO, "hello!");
+    /// span.follows_from(Span::current());
+    /// # }
+    /// ```
+    ///
+    /// Setting a `follows_from` relationship through a `&Span`:
+    /// ```
+    /// # use tracing::{span, Id, Level, Span};
+    /// # fn main() {
+    /// let span = span!(Level::INFO, "hello!");
+    /// let curr = Span::current();
+    /// span.follows_from(&curr);
     /// # }
     /// ```
     pub fn follows_from(&self, from: impl Into<Option<Id>>) -> &Self {

--- a/tracing/src/span.rs
+++ b/tracing/src/span.rs
@@ -761,16 +761,6 @@ impl Span {
     /// span.follows_from(id);
     /// # }
     /// ```
-    ///
-    /// `&Id` will not compile.
-    /// ```compile_fail
-    /// # use tracing::{span, Id, Level, Span};
-    /// # fn main() {
-    /// let span = span!(Level::INFO, "hello!");
-    /// let id = span.id();
-    /// span.follows_from(&id);
-    /// # }
-    /// ```
     pub fn follows_from(&self, from: impl Into<Option<Id>>) -> &Self {
         if let Some(ref inner) = self.inner {
             if let Some(from) = from.into() {

--- a/tracing/src/span.rs
+++ b/tracing/src/span.rs
@@ -720,10 +720,21 @@ impl Span {
     ///
     /// If this span is disabled, or the resulting follows-from relationship
     /// would be invalid, this function will do nothing.
-    pub fn follows_from(&self, from: impl for<'a> Into<Option<&'a Id>>) -> &Self {
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use tracing::{span, Id, Level, Span};
+    /// # fn main() {
+    /// let span = span!(Level::INFO, "hello!");
+    /// let curr = Span::current();
+    /// span.follows_from(curr.id());
+    /// # }
+    /// ```
+    pub fn follows_from(&self, from: impl Into<Option<Id>>) -> &Self {
         if let Some(ref inner) = self.inner {
             if let Some(from) = from.into() {
-                inner.follows_from(from);
+                inner.follows_from(&from);
             }
         }
         self

--- a/tracing/src/span.rs
+++ b/tracing/src/span.rs
@@ -723,17 +723,17 @@ impl Span {
     ///
     /// # Examples
     ///
-    /// Setting a `follows_from` relationship through an `Option<Id>`:
+    /// Setting a `follows_from` relationship with a `Span`:
     /// ```
     /// # use tracing::{span, Id, Level, Span};
     /// # fn main() {
-    /// let span = span!(Level::INFO, "hello!");
-    /// let curr = Span::current();
-    /// span.follows_from(curr.id());
+    /// let span1 = span!(Level::INFO, "span_1");
+    /// let span2 = span!(Level::DEBUG, "span_2");
+    /// span2.follows_from(span1);
     /// # }
     /// ```
     ///
-    /// Setting a `follows_from` relationship with a `Span`:
+    /// Setting a `follows_from` relationship with a `Span::current()`:
     /// ```
     /// # use tracing::{span, Id, Level, Span};
     /// # fn main() {
@@ -749,6 +749,36 @@ impl Span {
     /// let span = span!(Level::INFO, "hello!");
     /// let curr = Span::current();
     /// span.follows_from(&curr);
+    /// # }
+    /// ```
+    ///
+    /// Setting a `follows_from` relationship with an `Id`:
+    /// ```
+    /// # use tracing::{span, Id, Level, Span};
+    /// # fn main() {
+    /// let span = span!(Level::INFO, "hello!");
+    /// let id = span.id();
+    /// span.follows_from(id);
+    /// # }
+    /// ```
+    ///
+    /// `&Id` will not compile.
+    /// ```compile_fail
+    /// # use tracing::{span, Id, Level, Span};
+    /// # fn main() {
+    /// let span = span!(Level::INFO, "hello!");
+    /// let id = span.id();
+    /// span.follows_from(&id);
+    /// # }
+    /// ```
+    ///
+    /// Note that cloning an `Id` is cheap, so the following will compile:
+    /// ```
+    /// # use tracing::{span, Id, Level, Span};
+    /// # fn main() {
+    /// # let span = span!(Level::INFO, "hello!");
+    /// # let id = span.id();
+    /// span.follows_from(&id.clone());
     /// # }
     /// ```
     pub fn follows_from(&self, from: impl Into<Option<Id>>) -> &Self {

--- a/tracing/src/span.rs
+++ b/tracing/src/span.rs
@@ -771,16 +771,6 @@ impl Span {
     /// span.follows_from(&id);
     /// # }
     /// ```
-    ///
-    /// Note that cloning an `Id` is cheap, so the following will compile:
-    /// ```
-    /// # use tracing::{span, Id, Level, Span};
-    /// # fn main() {
-    /// # let span = span!(Level::INFO, "hello!");
-    /// # let id = span.id();
-    /// span.follows_from(&id.clone());
-    /// # }
-    /// ```
     pub fn follows_from(&self, from: impl Into<Option<Id>>) -> &Self {
         if let Some(ref inner) = self.inner {
             if let Some(from) = from.into() {

--- a/tracing/src/span.rs
+++ b/tracing/src/span.rs
@@ -733,7 +733,7 @@ impl Span {
     /// # }
     /// ```
     ///
-    /// Setting a `follows_from` relationship through a `Span`:
+    /// Setting a `follows_from` relationship with a `Span`:
     /// ```
     /// # use tracing::{span, Id, Level, Span};
     /// # fn main() {

